neqo-common/src/datagram.rs

```diff
@@ -44,17 +41,85 @@ impl Datagram {
     pub fn set_tos(&mut self, tos: IpTos) {
         self.tos = tos;
     }
+
+    #[must_use]
+    pub const fn segment_size(&self) -> usize {
+        self.segment_size
+    }
 }
 
-impl Deref for Datagram {
-    type Target = Vec<u8>;
+impl<D: AsRef<[u8]>> Datagram<D> {
+    pub fn new(
+        src: SocketAddr,
+        dst: SocketAddr,
+        tos: IpTos,
+        d: D,
+        segment_size: Option<usize>,
+    ) -> Self {
+        Self {
+            src,
+            dst,
+            tos,
+            segment_size: segment_size.unwrap_or_else(|| d.as_ref().len()),
+            d,
+        }
+    }
+    pub fn iter_segments(&self) -> impl Iterator<Item = &[u8]> {
+        self.d.as_ref().chunks(self.segment_size)
+    }
+
+    pub fn num_segments(&self) -> usize {
+        self.d.as_ref().len().div_ceil(self.segment_size)
+    }
+}
+
+impl<D: Deref<Target = [u8]>> Deref for Datagram<D> {
+    type Target = [u8];
     #[must_use]
     fn deref(&self) -> &Self::Target {
         &self.d
     }
 }
 
-impl std::fmt::Debug for Datagram {
+impl<'a> From<&'a Datagram> for Datagram<&'a [u8]> {
+    fn from(value: &'a Datagram) -> Self {
+        let Datagram {
+            src,
+            dst,
+            tos,
+            segment_size,
+            d,
+        } = value;
+        Datagram {
+            src: *src,
+            dst: *dst,
+            tos: *tos,
+            segment_size: *segment_size,
+            d,
+        }
+    }
+}
+
+impl From<Datagram<&[u8]>> for Datagram {
```

Comment:
Should this be a ToOwned implementation instead?

Reply:
I don't think implementing `ToOwned` is possible.

``` rust
pub trait ToOwned {
    type Owned: Borrow<Self>;

    fn to_owned(&self) -> Self::Owned;
}

```

https://doc.rust-lang.org/alloc/borrow/trait.ToOwned.html

`ToOwned` requires the associated type `Owned` to implement `Borrow<Self>`, in other words for `Datagram<Vec<u8>>` to implement `Borrow<Datagram<&'a [u8]>`. I don't think that is possible, see https://github.com/mozilla/neqo/pull/2093#discussion_r1788539953.